tests/test_repro.py

```diff
@@ -53,6 +54,79 @@ def test(self):
         self.assertNotEqual(ret, 0)
 
 
+class TestReproCyclicGraph(TestDvc):
+    def test(self):
+        self.dvc.run(deps=[self.FOO],
+                     outs=['bar.txt'],
+                     cmd='echo bar > bar.txt')
+
+        self.dvc.run(deps=['bar.txt'],
+                     outs=['baz.txt'],
+                     cmd='echo baz > baz.txt')
+
+        with open('cycle.dvc', 'w') as fd:
+            stage_dump = {
+                'cmd': 'echo baz > foo',
+                'deps': [{'path': 'baz.txt'}],
+                'outs': [{'path': self.FOO}],
+            }
+
+            yaml.safe_dump(stage_dump, fd, default_flow_style=False)
+
+        with self.assertRaises(CyclicGraphError):
+            self.dvc.reproduce('cycle.dvc')
+
+
+class TestReproWorkingDirectoryAsOutput(TestDvc):
+    def test(self):
+        os.mkdir(os.path.join(self.dvc.root_dir, 'dir1'))
+
+        self.dvc.run(
+            cwd='dir1',
+            outs=['../dir2'],
+            cmd='mkdir {path}'.format(path=os.path.join('..', 'dir2'))
+        )
+
+        faulty_stage_path = os.path.join('dir2', 'something.dvc')
+
+        with open(faulty_stage_path, 'w') as fd:
+            stage_dump = {
+                'cmd': 'echo something > something',
+                'outs': [{'path': 'something'}],
+            }
+
+            yaml.safe_dump(stage_dump, fd, default_flow_style=False)
+
+        with self.assertRaises(WorkingDirectoryAsOutputError):
+            self.dvc.reproduce(faulty_stage_path)
+
+    def test_nested(self):
+        os.mkdir(os.path.join(self.dvc.root_dir, 'dir1'))
+
+        self.dvc.run(
+            cwd='dir1',
+            outs=['../dir2'],
+            cmd='mkdir {path}'.format(path=os.path.join('..', 'dir2'))
+        )
+
+        nested_dir = os.path.join(self.dvc.root_dir, 'dir2', 'nested')
+
+        os.mkdir(nested_dir)
+
+        faulty_stage_path = os.path.join(nested_dir, 'error.dvc')
+
+        with open(faulty_stage_path, 'w') as fd:
+            stage_dump = {
+                'cmd': 'echo something > something',
+                'outs': [{'path': 'something'}],
+            }
+
+            yaml.safe_dump(stage_dump, fd, default_flow_style=False)
+
+        with self.assertRaises(WorkingDirectoryAsOutputError):
+            self.dvc.reproduce(faulty_stage_path)
```

Comment:
@efiop , any suggestions for this one?

```
======================================================================
ERROR: test_nested (tests.test_repro.TestReproWorkingDirectoryAsOutput)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\projects\dvc\tests\test_repro.py", line 127, in test_nested
    self.dvc.reproduce(faulty_stage_path)
  File "C:\projects\dvc\dvc\project.py", line 480, in reproduce
    ignore_build_cache=ignore_build_cache)
  File "C:\projects\dvc\dvc\project.py", line 509, in _reproduce
    ignore_build_cache)
  File "C:\projects\dvc\dvc\project.py", line 530, in _reproduce_stages
    for n in nx.dfs_postorder_nodes(G, node):
  File "c:\python27\lib\site-packages\networkx\algorithms\traversal\depth_first_search.py", line 273, in <genexpr>
    return (v for u, v, d in edges if d == 'reverse')
  File "c:\python27\lib\site-packages\networkx\algorithms\traversal\depth_first_search.py", line 405, in dfs_labeled_edges
    stack = [(start, depth_limit, iter(G[start]))]
  File "c:\python27\lib\site-packages\networkx\classes\graph.py", line 458, in __getitem__
    return self.adj[n]
  File "c:\python27\lib\site-packages\networkx\classes\coreviews.py", line 81, in __getitem__
    return AtlasView(self._atlas[name])
KeyError: 'dir2\\nested\\error.dvc'
```

Reply:
This error happens because `error.dvc` was not loaded into DAG, because it is out-of-scope. In order to test this, you should make sure that `error.dvc` get's loaded _before_ the first stage, because otherwise that directory would be ignored and error.dvc would be not discovered. See `stages()` method, where when we load some stage, we remember its outputs and exclude them from our search when walk()-ing the tree.

What we should do instead is something like:
```
        # NOTE: os.walk() walks in a sorted order and we need dir2 subdirs to
        # be processed before dir1 to load error.dvc first.                  
        dir1 = 'b'                                                           
        dir2 = 'a'                                                           
                                                                             
        os.mkdir(dir1)                                                       
        os.mkdir(dir2)                                                       
                                                                             
        nested_dir2 = os.path.join(dir2, 'nested2')                          
                                                                             
        out_dir2 = os.path.relpath(nested_dir2, dir1)                        
                                                                             
        self.dvc.run(                                                        
            cwd=dir1,                                                        
            outs=[out_dir2],                                                 
            cmd='mkdir {path}'.format(path=out_dir2)                         
        )                                                                    
                                                                             
        faulty_stage_path = os.path.join(nested_dir2, 'error.dvc')           
                                                                             
        with open(faulty_stage_path, 'w') as fd:                             
            stage_dump = {                                                   
                'cmd': 'echo something > something',                         
                'outs': [{'path': 'something'}],                             
            }                                                                
                                                                             
            yaml.safe_dump(stage_dump, fd, default_flow_style=False)         
                                                                             
        with self.assertRaises(WorkingDirectoryAsOutputError):               
            self.dvc.reproduce(faulty_stage_path)                            
```